src/tss2-fapi/ifapi_ima_eventlog.c

```diff
@@ -0,0 +1,853 @@
+    { .field_id = "n-ng",
+      .field_cb = eventname_ng_json_cb},
+    { .field_id = "sig",
+      .field_cb = signature_cb},
+};
+
+/**  Convert the IMA event to output format.
+ *
+ * Based on the event description in the ima template the appropriate callbacks are
+ * determined and executed to generate the ouput.
+ *
+ * @param[in] template The IMA event.
+ * @param[out] jso The output object.
+ * @param[out] name The name of the event object.
+ * @retval TSS2_RC_SUCCESS if the function call was a success.
+ * @retval TSS2_FAPI_RC_BAD_VALUE If the current template foramt can't be processed.
+ * @retval TSS2_FAPI_RC_MEMORY: if the FAPI cannot allocate enough memory.
+ */
+static TSS2_RC
+convert_ima_event_buffer(
+    IFAPI_IMA_TEMPLATE *template,
+    json_object *jso,
+    char **name)
+{
+    TSS2_RC r;
+    size_t offset = 0;
+    size_t i, j;
+    char *copy_of_template_format = NULL;
+    char *copy_of_template_format_orig = NULL;
+    char *current_field;
+    struct template_description *template_desc = NULL;
+    size_t size_of_template_tab = sizeof(template_tab) / sizeof(template_tab[0]);
+    size_t size_of_field_tab = sizeof(field_tab) / sizeof(field_tab[0]);
+    json_object *jso_content;
+
+    /* Search template decription corresponding to the template type. */
+    for (i = 0; i < size_of_template_tab; i++) {
+        if (strcmp(template->ima_type, template_tab[i].ima_type) == 0) {
+            template_desc = template_tab + i;
+            break;
+        }
+    }
+
+    if (i == size_of_template_tab) {
+        return_error(TSS2_FAPI_RC_BAD_VALUE, "Ima template type not supported.");
+    }
+
+    copy_of_template_format = strdup(template_desc->format);
+    /* Save pointer for cleanup. */
+    copy_of_template_format_orig = copy_of_template_format;
+    goto_if_null(copy_of_template_format, "Out of memory.", TSS2_FAPI_RC_MEMORY, error);
+    /* Loop over template fields separated by | */
+    for (i = 0; (current_field = strsep(&copy_of_template_format, "|")) != NULL; i++) {
+        struct template_field *field = NULL;
+
+        for (j = 0; j < size_of_field_tab; j++) {
+            if (!strcmp(current_field, field_tab[j].field_id)) {
+                field = field_tab + j;
+                break;
+            }
+        }
+
+        goto_if_null2(field, "Unknown field %s", r, TSS2_FAPI_RC_BAD_VALUE, error, current_field);
+
+        /* Convert the IMA data with the found callback. */
+        r = field->field_cb(&template->header.digest[0],
+                            template->event_buffer, &offset, jso, template);
+        *name= template->name;
+        goto_if_error(r, "Get field", error);
+    }
+    if (jso) {
+        r = get_json_content(jso, &jso_content);
+        return_if_error(r, "Get sub event");
+
+        r = add_uint8_ary_to_json(template->event_buffer, template->event_size, jso_content, "template_value");
+        return_if_error(r, "Create data to be hashed");
```

Comment:
shouldn't these 2 return_if_error be goto_if_error to free copy_of_template_format_orig?